server/services/invidious.ts

```diff
@@ -0,0 +1,426 @@
+			const bestMp4 = maxBy(mp4, score);
+			const bestAny = maxBy(list, score)!;
+			const chosen: InvidiousFormat = bestMp4 ?? bestAny;
+
+			// Infer MIME from container/type; Invidious does not always expose full codec/MIME details for YouTube progressive formats. – Invidious/YouTube issue
+			let mime =
+				chosen.type?.split(";")[0]?.trim() ||
+				(chosen.container === "webm"
+					? "video/webm"
+					: chosen.container === "mp4"
+					? "video/mp4"
+					: undefined);
+			// Always proxy via the instance (avoids CORS/signature churn) and carry itag if known.
+			const proxied = this.proxiedProgressive(host, id, chosen.itag);
+
+			// If MIME cannot be inferred reliably, prefer MP4 for widest browser compatibility. – Browser compatibility
+			if (!mime) {
+				const m =
+					typeof chosen.itag === "number" ? ITAG_TO_MIME.get(chosen.itag) : undefined;
+				mime = m ?? "video/mp4";
+			}
+			log.info("Chosen progressive stream via /latest_version", {
+				host,
+				id,
+				itag: chosen.itag,
+				label: chosen.qualityLabel,
+				container: chosen.container,
+				bitrate: chosen.bitrate,
+				mime,
+			});
+			return { service: "direct", id: proxied, ...base, mime };
+		}
+		// Fallback: let the instance pick a default via /latest_version (proxied)
+		const fallback = this.proxiedProgressive(host, id);
+		// Some videos only expose adaptive (DASH/HLS) and no progressive tracks. – YouTube encoding variability
+		log.warn("No progressive formatStreams found; falling back to proxied /latest_version", {
+			host,
+			id,
+		});
+		return {
+			service: "direct",
+			id: fallback,
+			...base,
+			mime: "video/mp4",
+		};
+	}
+	/** Probe a manifest and return its top bitrate/resolution.
+	 * Parse MPD/M3U8 using dedicated parsers; regex is unreliable because these manifests evolve frequently across YouTube encodes and Invidious rewrites. – YouTube/HLS/DASH variability
+	 */
+	private async probeManifest(
+		host: string,
+		id: string,
+		kind: "dash" | "hls"
+	): Promise<{ kind: "dash" | "hls"; url: string; topKbps: number; topRes?: string } | null> {
+		const url = this.manifestUrl(host, id, kind);
+		try {
+			const resp = await this.api.get<string>(url, {
+				headers: {
+					Accept:
+						kind === "dash"
+							? "application/dash+xml,text/plain;q=0.9,*/*;q=0.8"
+							: "application/vnd.apple.mpegurl,application/x-mpegURL,text/plain;q=0.9,*/*;q=0.8",
+				},
+				responseType: "text",
+				transformResponse: [r => r],
+			});
+			const text = typeof resp.data === "string" ? resp.data : String(resp.data);
+			if (kind === "dash") {
+				// 1) Faster, namespace- scan <Representation .../>-attr
+				let topBw = 0;
+				let topRes: string | undefined;
+
+				const repTagRe = /<Representation\b[^>]*\/?>/gi;
+				const attrNum = (src: string, name: string): number => {
+					// exakter Attributname (verhindert 'width' in 'bandwidth')
+					const re = new RegExp(`(?:^|\\s)${name}(?=\\s*=)\\s*=\\s*["'](\\d+)["']`, "i");
```

Comment:
It appears we've regressed here regarding regex instead of real parsing. I especially don't like the dynamic regex.

1. We shouldn't be building dynamic regexs
2. I realize I may have misled you a little [here](https://github.com/dyc3/opentogethertube/pull/1862#discussion_r2292278733). What I meant was that we already pull in the package, so we can just add it to our direct dependencies and be able to start using it. And we can just use the slightly outdated version without bloating `node_modules` (as long as there's no security advisories, of course).